File: kirun-js/src/engine/function/system/object/ObjectValues.ts
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';

import { isNullValue } from '../../../util/NullCheck';
import { EventResult } from '../../../model/EventResult';
import { AbstractObjectFunction } from './AbstractObjectFunction';
import { Schema } from '../../../json/schema/Schema';
import { duplicate } from '../../../util/duplicate';

const VALUE = 'value';

export class ObjectValues extends AbstractObjectFunction {
    public constructor() {
        super('ObjectValues', Schema.ofArray(VALUE, Schema.ofAny(VALUE)));
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        var source = context.getArguments()?.get('source');

        if (isNullValue(source))
            return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, []]]))]);

        let objectValues: any[] = Object.entries(duplicate(source))
            .sort((a, b) => a[0].localeCompare(b[0]))
            .map((e) => e[1]);

        return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, objectValues]]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/object/object_values.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.system.object.abstract_object_function import AbstractObjectFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.util.duplicate import duplicate
from kirun_py.util.null_check import is_null_value

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'


class ObjectValues(AbstractObjectFunction):

    def __init__(self) -> None:
        super().__init__('ObjectValues', Schema.of_array(VALUE, Schema.of_any(VALUE)))

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        source = context.get_arguments().get('source') if context.get_arguments() else None

        if is_null_value(source) or isinstance(source, (int, float, bool)):
            return FunctionOutput([EventResult.output_of({VALUE: []})])

        if isinstance(source, str):
            values = list(source)
            return FunctionOutput([EventResult.output_of({VALUE: values})])

        if isinstance(source, list):
            return FunctionOutput([EventResult.output_of({VALUE: list(source)})])

        if isinstance(source, dict):
            object_values = [v for _, v in sorted(source.items())]
            return FunctionOutput([EventResult.output_of({VALUE: object_values})])

        return FunctionOutput([EventResult.output_of({VALUE: []})])
